File: bug67_duplicate_processing_in_loop/bug67_duplicate_processing_in_loop/main.js
function main() {
    let data = [];
    for (let i = 0n; i < 10n; i++) data.push(0x100n * (i + 1n));
    
    let acc = 0n;
    for (let j = 0; j < 10; j++) {
        let i = BigInt(j);
        let v = data[j];
        acc ^= v * 0x9E3779B97F4A7C15n;
        acc &= 0xFFFFFFFFFFFFFFFFn;
        acc ^= v * 0x9E3779B97F4A7C15n;
        acc &= 0xFFFFFFFFFFFFFFFFn;
        acc = ((acc << (i % 9n + 1n)) | (acc >> (63n - i % 9n))) & 0xFFFFFFFFFFFFFFFFn;
    }
    
    let correct = 0n;
    for (let j = 0; j < 10; j++) {
        let i = BigInt(j);
        let v = data[j];
        correct ^= v * 0x9E3779B97F4A7C15n;
        correct &= 0xFFFFFFFFFFFFFFFFn;
        correct = ((correct << (i % 9n + 1n)) | (correct >> (63n - i % 9n))) & 0xFFFFFFFFFFFFFFFFn;
    }
    
    if (acc === correct) {
        console.error("AssertionError: duplicate xor cancels and same result?");
        process.exit(1);
    }
    console.error(`ValueError: xor applied twice cancels itself: got 0x${acc.toString(16)} expected 0x${correct.toString(16)}`);
    process.exit(1);
}
main();

File: bug67_duplicate_processing_in_loop/bug67_duplicate_processing_in_loop/main.c
#include <stdio.h>

int main() {
    int data[10];
    for (int i = 0; i < 10; i++) data[i] = 0x100 * (i + 1);
    
    unsigned long long acc = 0;
    for (int i = 0; i < 10; i++) {
        acc ^= (unsigned long long)data[i] * 0x9E3779B97F4A7C15ULL;
        acc ^= (unsigned long long)data[i] * 0x9E3779B97F4A7C15ULL;
        acc = ((acc << (i % 9 + 1)) | (acc >> (63 - i % 9)));
    }
    
    unsigned long long correct = 0;
    for (int i = 0; i < 10; i++) {
        correct ^= (unsigned long long)data[i] * 0x9E3779B97F4A7C15ULL;
        correct = ((correct << (i % 9 + 1)) | (correct >> (63 - i % 9)));
    }
    
    if (acc == correct) {
        fprintf(stderr, "AssertionError: duplicate xor cancels and same result?\n");
        return 1;
    }
    
    fprintf(stderr, "ValueError: xor applied twice cancels itself: got 0x%llx expected 0x%llx\n", acc, correct);
    return 1;
}
